File: agent_slack/coordination/handoff.py
"""
Handoff Primitive

Handoffs are how agents pass work or context to each other.
They're more structured than just posting a message - they include
context, expectations, and acknowledgment.
"""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, Any
from enum import Enum
import uuid


class HandoffReason(Enum):
    """Why work is being handed off."""
    
    # Capability-based
    CAPABILITY_NEEDED = "capability_needed"  # Need a capability I don't have
    BETTER_SUITED = "better_suited"          # Another agent is better for this
    
    # Workload-based
    AT_CAPACITY = "at_capacity"              # I'm too busy
    LOAD_BALANCING = "load_balancing"        # Distributing work
    
    # Process-based
    NEXT_STEP = "next_step"                  # Moving to next phase
    REVIEW_NEEDED = "review_needed"          # Need someone to review
    APPROVAL_NEEDED = "approval_needed"      # Need approval to continue
    
    # Issue-based
    STUCK = "stuck"                          # Can't make progress
    ERROR = "error"                          # Hit an error
    ESCALATION = "escalation"                # Needs higher-level attention
    
    # Completion
    COMPLETED = "completed"                  # Passing back completed work


class HandoffStatus(Enum):
    """Status of a handoff."""
    
    PENDING = "pending"          # Created, waiting for acceptance
    ACCEPTED = "accepted"        # Recipient accepted
    REJECTED = "rejected"        # Recipient rejected
    IN_PROGRESS = "in_progress"  # Being processed
    COMPLETED = "completed"      # Successfully handled
    CANCELLED = "cancelled"      # Cancelled by sender
    EXPIRED = "expired"          # Timed out


@dataclass
class HandoffContext:
    """
    Context being passed in a handoff.
    
    This bundles everything the receiving agent needs to continue the work.
    """
    
    # Summary for quick understanding
    summary: str = ""
    
    # Detailed context
    background: str = ""          # What led to this point
    current_state: str = ""       # Where things stand now
    expected_outcome: str = ""    # What success looks like
    
    # Linked resources
    message_ids: list[str] = field(default_factory=list)
    thread_ids: list[str] = field(default_factory=list)
    artifact_ids: list[str] = field(default_factory=list)
    task_ids: list[str] = field(default_factory=list)
    
    # Structured data
    data: dict = field(default_factory=dict)
    
    def to_dict(self) -> dict:
        return {
            "summary": self.summary,
            "background": self.background,
            "current_state": self.current_state,
            "expected_outcome": self.expected_outcome,
            "message_ids": self.message_ids,
            "thread_ids": self.thread_ids,
            "artifact_ids": self.artifact_ids,
            "task_ids": self.task_ids,
            "data": self.data,
        }


@dataclass
class Handoff:
    """
    A handoff of work from one agent to another.
    
    Key concepts:
    - Handoffs are explicit (not just hoping someone picks up a message)
    - They include context so the receiver can continue effectively
    - They require acknowledgment (accept/reject)
    - They track completion status
    
    Design decisions:
    - Handoffs are first-class objects, not just message metadata
    - Context is structured for machine consumption
    - Handoffs can be to a specific agent or to any agent with capability
    - Expiration prevents stale handoffs
    """
    
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    
    # Participants
    from_agent: str = ""                  # Who is handing off
    to_agent: Optional[str] = None        # Specific recipient (or None for open)
    to_capability: Optional[str] = None   # Required capability (for open handoffs)
    
    # Reason and context
    reason: HandoffReason = HandoffReason.NEXT_STEP
    context: HandoffContext = field(default_factory=HandoffContext)
    
    # Status
    status: HandoffStatus = HandoffStatus.PENDING
    
    # Location
    channel_id: Optional[str] = None
    thread_id: Optional[str] = None       # Handoff discussion thread
    
    # Timing
    created_at: datetime = field(default_factory=datetime.utcnow)
    accepted_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    
    # Response
    accepted_by: Optional[str] = None     # Who accepted (may differ from to_agent)
    rejection_reason: str = ""
    completion_notes: str = ""
    
    # Priority
    is_urgent: bool = False
    
    @property
    def is_open(self) -> bool:
        """Check if handoff is still waiting for action."""
        return self.status in (HandoffStatus.PENDING, HandoffStatus.ACCEPTED, HandoffStatus.IN_PROGRESS)
    
    @property
    def is_targeted(self) -> bool:
        """Check if handoff is to a specific agent."""
        return self.to_agent is not None
    
    @property
    def is_expired(self) -> bool:
        """Check if handoff has expired."""
        if self.expires_at and datetime.utcnow() > self.expires_at:
            return True
        return self.status == HandoffStatus.EXPIRED
    
    def accept(self, agent_id: str) -> "Handoff":
        """Accept the handoff."""
        self.status = HandoffStatus.ACCEPTED
        self.accepted_by = agent_id
        self.accepted_at = datetime.utcnow()
        return self
    
    def reject(self, agent_id: str, reason: str = "") -> "Handoff":
        """Reject the handoff."""
        self.status = HandoffStatus.REJECTED
        self.rejection_reason = reason
        return self
    
    def start(self) -> "Handoff":
        """Mark handoff as being worked on."""
        self.status = HandoffStatus.IN_PROGRESS
        return self
    
    def complete(self, notes: str = "") -> "Handoff":
        """Mark handoff as completed."""
        self.status = HandoffStatus.COMPLETED
        self.completed_at = datetime.utcnow()
        self.completion_notes = notes
        return self
    
    def cancel(self) -> "Handoff":
        """Cancel the handoff."""
        self.status = HandoffStatus.CANCELLED
        return self
    
    def expire(self) -> "Handoff":
        """Mark handoff as expired."""
        self.status = HandoffStatus.EXPIRED
        return self
    
    def to_dict(self) -> dict:
        """Serialize to dictionary."""
        return {
            "id": self.id,
            "from_agent": self.from_agent,
            "to_agent": self.to_agent,
            "to_capability": self.to_capability,
            "reason": self.reason.value,
            "status": self.status.value,
            "context": self.context.to_dict(),
            "channel_id": self.channel_id,
            "thread_id": self.thread_id,
            "is_urgent": self.is_urgent,
            "is_open": self.is_open,
            "is_targeted": self.is_targeted,
            "created_at": self.created_at.isoformat(),
            "accepted_at": self.accepted_at.isoformat() if self.accepted_at else None,
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
            "accepted_by": self.accepted_by,
        }
    
    def __str__(self) -> str:
        status_icons = {
            HandoffStatus.PENDING: "⏳",
            HandoffStatus.ACCEPTED: "🤝",
            HandoffStatus.REJECTED: "❌",
            HandoffStatus.IN_PROGRESS: "🔄",
            HandoffStatus.COMPLETED: "✅",
            HandoffStatus.CANCELLED: "🚫",
            HandoffStatus.EXPIRED: "⏰",
        }
        icon = status_icons.get(self.status, "❓")
        target = self.to_agent or f"[{self.to_capability}]"
        urgent = "🚨 " if self.is_urgent else ""
        return f"{urgent}{icon} {self.from_agent[:8]} -> {target}: {self.reason.value}"


class HandoffManager:
    """
    Manages handoffs within a workspace.
    """
    
    def __init__(self):
        self._handoffs: dict[str, Handoff] = {}
        self._by_sender: dict[str, list[str]] = {}
        self._by_recipient: dict[str, list[str]] = {}
        self._pending_by_capability: dict[str, list[str]] = {}
    
    def create(
        self,
        from_agent: str,
        reason: HandoffReason,
        context: HandoffContext,
        to_agent: Optional[str] = None,
        to_capability: Optional[str] = None,
        channel_id: Optional[str] = None,
        is_urgent: bool = False,
    ) -> Handoff:
        """Create a new handoff."""
        if not to_agent and not to_capability:
            raise ValueError("Must specify either to_agent or to_capability")
        
        handoff = Handoff(
            from_agent=from_agent,
            to_agent=to_agent,
            to_capability=to_capability,
            reason=reason,
            context=context,
            channel_id=channel_id,
            is_urgent=is_urgent,
        )
        
        return self.add(handoff)
    
    def add(self, handoff: Handoff) -> Handoff:
        """Add a handoff to the manager."""
        self._handoffs[handoff.id] = handoff
        
        # Index by sender
        if handoff.from_agent not in self._by_sender:
            self._by_sender[handoff.from_agent] = []
        self._by_sender[handoff.from_agent].append(handoff.id)
        
        # Index by recipient
        if handoff.to_agent:
            if handoff.to_agent not in self._by_recipient:
                self._by_recipient[handoff.to_agent] = []
            self._by_recipient[handoff.to_agent].append(handoff.id)
        
        # Index by capability (for open handoffs)
        if handoff.to_capability:
            if handoff.to_capability not in self._pending_by_capability:
                self._pending_by_capability[handoff.to_capability] = []
            self._pending_by_capability[handoff.to_capability].append(handoff.id)
        
        return handoff
    
    def get(self, handoff_id: str) -> Optional[Handoff]:
        """Get a handoff by ID."""
        return self._handoffs.get(handoff_id)
    
    def get_pending_for_agent(self, agent_id: str) -> list[Handoff]:
        """Get pending handoffs targeted at an agent."""
        handoff_ids = self._by_recipient.get(agent_id, [])
        handoffs = [self._handoffs[id] for id in handoff_ids if id in self._handoffs]
        return [h for h in handoffs if h.status == HandoffStatus.PENDING]
    
    def get_pending_for_capability(self, capability: str) -> list[Handoff]:
        """Get pending handoffs for a capability."""
        handoff_ids = self._pending_by_capability.get(capability, [])
        handoffs = [self._handoffs[id] for id in handoff_ids if id in self._handoffs]
        return [h for h in handoffs if h.status == HandoffStatus.PENDING]
    
    def get_sent_by(self, agent_id: str, open_only: bool = True) -> list[Handoff]:
        """Get handoffs sent by an agent."""
        handoff_ids = self._by_sender.get(agent_id, [])
        handoffs = [self._handoffs[id] for id in handoff_ids if id in self._handoffs]
        
        if open_only:
            handoffs = [h for h in handoffs if h.is_open]
        
        return sorted(handoffs, key=lambda h: h.created_at, reverse=True)
    
    def accept(self, handoff_id: str, agent_id: str) -> Optional[Handoff]:
        """Accept a handoff."""
        handoff = self.get(handoff_id)
        if not handoff or handoff.status != HandoffStatus.PENDING:
            return None
        
        handoff.accept(agent_id)
        
        # Update recipient index
        if agent_id not in self._by_recipient:
            self._by_recipient[agent_id] = []
        self._by_recipient[agent_id].append(handoff_id)
        
        # Remove from capability index
        if handoff.to_capability and handoff.to_capability in self._pending_by_capability:
            if handoff_id in self._pending_by_capability[handoff.to_capability]:
                self._pending_by_capability[handoff.to_capability].remove(handoff_id)
        
        return handoff
    
    def reject(
        self,
        handoff_id: str,
        agent_id: str,
        reason: str = "",
    ) -> Optional[Handoff]:
        """Reject a handoff."""
        handoff = self.get(handoff_id)
        if not handoff or handoff.status != HandoffStatus.PENDING:
            return None
        
        handoff.reject(agent_id, reason)
        return handoff
    
    def complete(self, handoff_id: str, notes: str = "") -> Optional[Handoff]:
        """Complete a handoff."""
        handoff = self.get(handoff_id)
        if not handoff:
            return None
        
        handoff.complete(notes)
        return handoff
    
    def cleanup_expired(self) -> list[str]:
        """Mark expired handoffs."""
        expired = []
        now = datetime.utcnow()
        
        for handoff in self._handoffs.values():
            if handoff.is_open and handoff.expires_at and now > handoff.expires_at:
                handoff.expire()
                expired.append(handoff.id)
        
        return expired
    
    def get_stats(self) -> dict:
        """Get handoff statistics."""
        all_handoffs = list(self._handoffs.values())
        
        return {
            "total": len(all_handoffs),
            "pending": len([h for h in all_handoffs if h.status == HandoffStatus.PENDING]),
            "in_progress": len([h for h in all_handoffs if h.status == HandoffStatus.IN_PROGRESS]),
            "completed": len([h for h in all_handoffs if h.status == HandoffStatus.COMPLETED]),
            "rejected": len([h for h in all_handoffs if h.status == HandoffStatus.REJECTED]),
            "by_reason": {
                r.value: len([h for h in all_handoffs if h.reason == r])
                for r in HandoffReason
            },
        }

File: src/coordination/handoff.ts
/**
 * Handoff Primitive
 *
 * Handoffs are how agents pass work or context to each other.
 * They're more structured than just posting a message - they include
 * context, expectations, and acknowledgment.
 */

import { z } from 'zod';
import { nanoid } from 'nanoid';

/**
 * Why work is being handed off.
 */
export const HandoffReason = {
  // Capability-based
  CAPABILITY_NEEDED: 'capability_needed',
  BETTER_SUITED: 'better_suited',

  // Workload-based
  AT_CAPACITY: 'at_capacity',
  LOAD_BALANCING: 'load_balancing',

  // Process-based
  NEXT_STEP: 'next_step',
  REVIEW_NEEDED: 'review_needed',
  APPROVAL_NEEDED: 'approval_needed',

  // Issue-based
  STUCK: 'stuck',
  ERROR: 'error',
  ESCALATION: 'escalation',

  // Completion
  COMPLETED: 'completed',
} as const;

export type HandoffReason = (typeof HandoffReason)[keyof typeof HandoffReason];

/**
 * Status of a handoff.
 */
export const HandoffStatus = {
  PENDING: 'pending',
  ACCEPTED: 'accepted',
  REJECTED: 'rejected',
  IN_PROGRESS: 'in_progress',
  COMPLETED: 'completed',
  CANCELLED: 'cancelled',
  EXPIRED: 'expired',
} as const;

export type HandoffStatus = (typeof HandoffStatus)[keyof typeof HandoffStatus];

/**
 * Context being passed in a handoff.
 */
export interface HandoffContext {
  summary: string;
  background?: string;
  currentState?: string;
  expectedOutcome?: string;
  messageIds?: string[];
  threadIds?: string[];
  artifactIds?: string[];
  taskIds?: string[];
  data?: Record<string, unknown>;
}

export const HandoffSchema = z.object({
  id: z.string(),
  fromAgent: z.string(),
  toAgent: z.string().optional(),
  toCapability: z.string().optional(),
  reason: z.enum([
    'capability_needed', 'better_suited', 'at_capacity', 'load_balancing',
    'next_step', 'review_needed', 'approval_needed',
    'stuck', 'error', 'escalation', 'completed',
  ]),
  status: z.enum(['pending', 'accepted', 'rejected', 'in_progress', 'completed', 'cancelled', 'expired']),
  channelId: z.string().optional(),
  threadId: z.string().optional(),
  isUrgent: z.boolean().default(false),
  createdAt: z.date(),
  acceptedAt: z.date().optional(),
  completedAt: z.date().optional(),
  expiresAt: z.date().optional(),
  acceptedBy: z.string().optional(),
  rejectionReason: z.string().default(''),
  completionNotes: z.string().default(''),
});

export type HandoffData = z.infer<typeof HandoffSchema>;

/**
 * A handoff of work from one agent to another.
 */
export class Handoff {
  readonly id: string;

  // Participants
  readonly fromAgent: string;
  readonly toAgent?: string;
  readonly toCapability?: string;

  // Reason and context
  readonly reason: HandoffReason;
  readonly context: HandoffContext;

  // Status
  status: HandoffStatus;

  // Location
  channelId?: string;
  threadId?: string;

  // Timing
  readonly createdAt: Date;
  acceptedAt?: Date;
  completedAt?: Date;
  expiresAt?: Date;

  // Response
  acceptedBy?: string;
  rejectionReason: string;
  completionNotes: string;

  // Priority
  readonly isUrgent: boolean;

  constructor(data: {
    fromAgent: string;
    reason: HandoffReason;
    context: HandoffContext;
    toAgent?: string;
    toCapability?: string;
    channelId?: string;
    isUrgent?: boolean;
    expiresAt?: Date;
  }) {
    if (!data.toAgent && !data.toCapability) {
      throw new Error('Must specify either toAgent or toCapability');
    }

    this.id = nanoid();
    this.fromAgent = data.fromAgent;
    this.toAgent = data.toAgent;
    this.toCapability = data.toCapability;
    this.reason = data.reason;
    this.context = data.context;
    this.status = HandoffStatus.PENDING;
    this.channelId = data.channelId;
    this.createdAt = new Date();
    this.expiresAt = data.expiresAt;
    this.isUrgent = data.isUrgent ?? false;
    this.rejectionReason = '';
    this.completionNotes = '';
  }

  get isOpen(): boolean {
    return (
      this.status === HandoffStatus.PENDING ||
      this.status === HandoffStatus.ACCEPTED ||
      this.status === HandoffStatus.IN_PROGRESS
    );
  }

  get isTargeted(): boolean {
    return this.toAgent !== undefined;
  }

  get isExpired(): boolean {
    if (this.expiresAt && new Date() > this.expiresAt) {
      return true;
    }
    return this.status === HandoffStatus.EXPIRED;
  }

  accept(agentId: string): this {
    this.status = HandoffStatus.ACCEPTED;
    this.acceptedBy = agentId;
    this.acceptedAt = new Date();
    return this;
  }

  reject(_agentId: string, reason = ''): this {
    this.status = HandoffStatus.REJECTED;
    this.rejectionReason = reason;
    return this;
  }

  start(): this {
    this.status = HandoffStatus.IN_PROGRESS;
    return this;
  }

  complete(notes = ''): this {
    this.status = HandoffStatus.COMPLETED;
    this.completedAt = new Date();
    this.completionNotes = notes;
    return this;
  }

  cancel(): this {
    this.status = HandoffStatus.CANCELLED;
    return this;
  }

  expire(): this {
    this.status = HandoffStatus.EXPIRED;
    return this;
  }

  toJSON(): HandoffData & { context: HandoffContext; isOpen: boolean; isTargeted: boolean } {
    return {
      id: this.id,
      fromAgent: this.fromAgent,
      toAgent: this.toAgent,
      toCapability: this.toCapability,
      reason: this.reason,
      status: this.status,
      context: this.context,
      channelId: this.channelId,
      threadId: this.threadId,
      isUrgent: this.isUrgent,
      isOpen: this.isOpen,
      isTargeted: this.isTargeted,
      createdAt: this.createdAt,
      acceptedAt: this.acceptedAt,
      completedAt: this.completedAt,
      expiresAt: this.expiresAt,
      acceptedBy: this.acceptedBy,
      rejectionReason: this.rejectionReason,
      completionNotes: this.completionNotes,
    };
  }

  toString(): string {
    const statusIcons: Record<HandoffStatus, string> = {
      pending: '⏳',
      accepted: '🤝',
      rejected: '❌',
      in_progress: '🔄',
      completed: '✅',
      cancelled: '🚫',
      expired: '⏰',
    };
    const icon = statusIcons[this.status];
    const target = this.toAgent ?? `[${this.toCapability}]`;
    const urgent = this.isUrgent ? '🚨 ' : '';
    return `${urgent}${icon} ${this.fromAgent.slice(0, 8)} -> ${target}: ${this.reason}`;
  }
}

/**
 * Manages handoffs within a workspace.
 */
export class HandoffManager {
  private readonly _handoffs: Map<string, Handoff> = new Map();
  private readonly _bySender: Map<string, string[]> = new Map();
  private readonly _byRecipient: Map<string, string[]> = new Map();
  private readonly _pendingByCapability: Map<string, string[]> = new Map();

  /**
   * Create a new handoff.
   */
  create(data: {
    fromAgent: string;
    reason: HandoffReason;
    context: HandoffContext;
    toAgent?: string;
    toCapability?: string;
    channelId?: string;
    isUrgent?: boolean;
    expiresAt?: Date;
  }): Handoff {
    const handoff = new Handoff(data);
    return this.add(handoff);
  }

  /**
   * Add a handoff to the manager.
   */
  add(handoff: Handoff): Handoff {
    this._handoffs.set(handoff.id, handoff);

    // Index by sender
    if (!this._bySender.has(handoff.fromAgent)) {
      this._bySender.set(handoff.fromAgent, []);
    }
    this._bySender.get(handoff.fromAgent)!.push(handoff.id);

    // Index by recipient
    if (handoff.toAgent) {
      if (!this._byRecipient.has(handoff.toAgent)) {
        this._byRecipient.set(handoff.toAgent, []);
      }
      this._byRecipient.get(handoff.toAgent)!.push(handoff.id);
    }

    // Index by capability (for open handoffs)
    if (handoff.toCapability) {
      if (!this._pendingByCapability.has(handoff.toCapability)) {
        this._pendingByCapability.set(handoff.toCapability, []);
      }
      this._pendingByCapability.get(handoff.toCapability)!.push(handoff.id);
    }

    return handoff;
  }

  /**
   * Get a handoff by ID.
   */
  get(handoffId: string): Handoff | undefined {
    return this._handoffs.get(handoffId);
  }

  /**
   * Get pending handoffs targeted at an agent.
   */
  getPendingForAgent(agentId: string): Handoff[] {
    const handoffIds = this._byRecipient.get(agentId) ?? [];
    return handoffIds
      .map((id) => this._handoffs.get(id)!)
      .filter((h) => h && h.status === HandoffStatus.PENDING);
  }

  /**
   * Get pending handoffs for a capability.
   */
  getPendingForCapability(capability: string): Handoff[] {
    const handoffIds = this._pendingByCapability.get(capability) ?? [];
    return handoffIds
      .map((id) => this._handoffs.get(id)!)
      .filter((h) => h && h.status === HandoffStatus.PENDING);
  }

  /**
   * Get handoffs sent by an agent.
   */
  getSentBy(agentId: string, options: { openOnly?: boolean } = {}): Handoff[] {
    const handoffIds = this._bySender.get(agentId) ?? [];
    let handoffs = handoffIds.map((id) => this._handoffs.get(id)!).filter(Boolean);

    if (options.openOnly !== false) {
      handoffs = handoffs.filter((h) => h.isOpen);
    }

    return handoffs.sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());
  }

  /**
   * Accept a handoff.
   */
  accept(handoffId: string, agentId: string): Handoff | undefined {
    const handoff = this.get(handoffId);
    if (!handoff || handoff.status !== HandoffStatus.PENDING) {
      return undefined;
    }

    handoff.accept(agentId);

    // Update recipient index
    if (!this._byRecipient.has(agentId)) {
      this._byRecipient.set(agentId, []);
    }
    this._byRecipient.get(agentId)!.push(handoffId);

    // Remove from capability index
    if (handoff.toCapability && this._pendingByCapability.has(handoff.toCapability)) {
      const ids = this._pendingByCapability.get(handoff.toCapability)!;
      const idx = ids.indexOf(handoffId);
      if (idx !== -1) ids.splice(idx, 1);
    }

    return handoff;
  }

  /**
   * Reject a handoff.
   */
  reject(handoffId: string, agentId: string, reason = ''): Handoff | undefined {
    const handoff = this.get(handoffId);
    if (!handoff || handoff.status !== HandoffStatus.PENDING) {
      return undefined;
    }

    handoff.reject(agentId, reason);
    return handoff;
  }

  /**
   * Complete a handoff.
   */
  complete(handoffId: string, notes = ''): Handoff | undefined {
    const handoff = this.get(handoffId);
    if (!handoff) return undefined;

    handoff.complete(notes);
    return handoff;
  }

  /**
   * Mark expired handoffs.
   */
  cleanupExpired(): string[] {
    const expired: string[] = [];
    const now = new Date();

    for (const handoff of this._handoffs.values()) {
      if (handoff.isOpen && handoff.expiresAt && now > handoff.expiresAt) {
        handoff.expire();
        expired.push(handoff.id);
      }
    }

    return expired;
  }

  /**
   * Get handoff statistics.
   */
  getStats(): {
    total: number;
    pending: number;
    inProgress: number;
    completed: number;
    rejected: number;
    byReason: Record<HandoffReason, number>;
  } {
    const allHandoffs = Array.from(this._handoffs.values());

    const byReason = {} as Record<HandoffReason, number>;
    for (const reason of Object.values(HandoffReason)) {
      byReason[reason] = allHandoffs.filter((h) => h.reason === reason).length;
    }

    return {
      total: allHandoffs.length,
      pending: allHandoffs.filter((h) => h.status === HandoffStatus.PENDING).length,
      inProgress: allHandoffs.filter((h) => h.status === HandoffStatus.IN_PROGRESS).length,
      completed: allHandoffs.filter((h) => h.status === HandoffStatus.COMPLETED).length,
      rejected: allHandoffs.filter((h) => h.status === HandoffStatus.REJECTED).length,
      byReason,
    };
  }
}
